message-scroll: Mark read for events without previous message ID.

laurynmm opened the pull request:
Previously, the event handler for marking selected messages as read was only triggered if there was a previously selected message ID. This meant that user navigation (either via hotkeys or mouse clicks) would not necessarily mark the initially selected first unread message in the view, even if the view/user settings allowed for the message to be marked read.

Occasionally, a document on scroll event would be called that would create a race where the message would get marked as read in cases where message fetch requests were received more slowly. This made the overall mark as read experience inconsistent.

We update the event handler to allow for marking the selected message as read when there was no previously selected message in the message list.

There is an initial loading of messages to the home message list where we set `mark_read: false` since we can't safely assume on a page load/reload that the home view is the user's current or default view. Once the home view message list is loaded, we add a check for the current view and mark the selected message as read if the home view is indeed the current view.

Fixes #26021.

See [relevant CZO conversation](https://chat.zulip.org/#narrow/stream/6-frontend/topic/inconsistent.20mark.20as.20read/near/1580820).

**Notes**:
- There's an issue with these changes in the case that a user has a unfocused browser tab open on a stream/topic view that's empty (aka there are no messages). While they are working elsewhere in the browser, someone else sends a new message to that same stream/topic. As this is the first message in the view and there is no previous selected message ID, the new message will be marked as read even though the browser tab is not focused. And, if the user doesn't go back to the tab before ending their browser session, then they may not notice the message was received and marked as read.
- Something noted in manual testing of these changes is that for users with All messages as their default view, if they logout from another view with unread messages, then the unread count will flicker down before the page loads to the log-in page. This is not due to these changes and is discussed [here on CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/logout.20hashchange/near/1591366).

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented on web/src/server_events.js:
```diff
@@ -264,7 +265,14 @@ export function force_get_events() {
 export function home_view_loaded() {
     waiting_on_homeview_load = false;
     get_events_success([]);
-    $(document).trigger("home_view_loaded.zulip");
```
This seems like an old event trigger that no longer does anything. A git-grep search for `home_view_loaded.zulip` only returns this.

timabbott replied:
`git log -Shome_view_loaded.zulip` would be a good way to check this.

laurynmm replied:
That was very helpful! Seems like the event code was removed in commit 92ad98814429c8854c725278c55bf976b927ad5c.

laurynmm replied:
This was taken care of in #26122.

laurynmm commented on web/src/message_scroll.js:
```diff
@@ -150,13 +150,18 @@ export function initialize() {
         }
 
-        if (event.mark_read && event.previously_selected_id !== -1) {
-            // Mark messages between old pointer and new pointer as read
```
I changed the comments here to refer to the selected message and not pointer as that doesn't seem to be a term we use in the code base much anymore.

timabbott replied:
Yeah, the pointer used to be a thing in the API -- and was important before we tracked unread messages individually, but was a bad model and has been mostly purged.

laurynmm commented:
@amanagr - Would you be up for looking over these changes when you get a chance?

amanagr commented:
I think this PR makes changes in the correct direction. Changes took good.

laurynmm commented on web/tests/server_events.test.js:
```diff
@@ -9,13 +9,6 @@ const {page_params} = require("./lib/zpage_params");
-            trigger() {},
-        };
-    },
-});
```
This was taken care of in #26122.

laurynmm commented:
**Updates**:
- I added a commit to deal with not marking the first selected message as read in an unfocused / not active session.
  - Situation: The user opens a stream/topic for a live chat that's starting soon. The user leaves the computer to go make coffee and gets a call while they are doing that. This means that they get back to their computer after the chat has begun.
  - We don't want any messages marked as read while they are away, which previously was (not) an issue as a view without a previously selected message ID wouldn't cause anything to be marked as read.

- When testing these changes, I noted that the message view might auto-scroll into a situation where the initial message received in the view is partially covered by the message header bar.
  - I've opened #26143 to address the issue of the message being auto-scrolled underneath the message header.

laurynmm commented on web/src/unread_ops.js:
```diff
@@ -484,6 +484,22 @@ export function initialize() {
+            // focus.
+            if (
+                narrow_state.is_message_feed_visible() &&
+                message_lists.current.selected_id() !== -1
```
For this case, where we want the selected message in the message list view potentially marked as read, I didn't add a check for an open modal, which happens in `process_visible`.

My thought process was that we want the selected message marked as read (if possible) when the modal is closed.

If the user closes the tab, without looking at it, then nothing is marked as read. If they user clicks into the tab, but closes the tab without closing the modal, then they could potentially miss that the selected message was marked as read.

laurynmm replied:
This comment is out of date with the current version of these changes. I decided that it was better to be consistent with `process_visible` and not mark as read unless `viewport_is_visible_and_focused` returns true.

That way there is a bug that is consistent with both cases (bottom of the feed is visible, bottom of the feed isn't visible) where a overlay or modal is open in an unfocused window and new messages are received in the view.

timabbott commented on web/src/message_fetch.js:
```diff
@@ -452,8 +452,11 @@ export function initialize(home_view_loaded) {
+            // user may not have the home view as their current view
+            // (for reloads) or their default view (for initial load).
             message_lists.home.select_id(data.anchor, {
+                mark_read: false,
```
Ooh, I suspect this change explains and fixes this longstanding bug: #19329 / #16986.

laurynmm replied:
So, a potential reproducer for those issues on `main`:
- Mark x number of messages unread for Cordelia (can me a mix of direct messages and stream messages)
- Open a tab for her in **All messages** view (doesn't have to be the default view) and pin it (I used Firefox in Linux). Make sure that the pinned tab is such that messages can be marked as read.
- Close Firefox
- Open Firefox

I see the pinned tab with the initial unread count I expect, and then after a few seconds of the tab reloading, it will go down by one. Note that I have no other browsers/tabs open with Zulip Dev.

And when on the branch with these changes, the above does not happen. The unread count won't go down until I click on the tab. It might be the fix here and it might be the fix with the window.focus in the second commit.

laurynmm replied:
Mmmm... that reproduces on `main` in a stream view with unread messages, so it's a bug but probably not fixed by this change.

timabbott replied:
Cool, that's a great reproducer -- I guess in particular what's happening there is the Firefox tab pinning behavior is a way that you have a noninteractive reload via a different code path.

And now when I read the logic at the top of `activity.js`, at least the logic there is notably not considering this possibility. That logic is I think not actually involved in marking things as read, but suggests we might have a broader bug here.

```
// When you open Zulip in a new browser window, client_is_active                                        
// should be true.  When a server-initiated reload happens, however,                                    
// it should be initialized to false.  We handle this with a check for                                  
// whether the window is focused at initialization time.                                                
export let client_is_active = document.hasFocus && document.hasFocus();         

// new_user_input is a more strict version of client_is_active used                                     
// primarily for analytics.  We initialize this to true, to count new                                   
// page loads, but set it to false in the onload function in reload.js                                  
// if this was a server-initiated-reload to avoid counting a                                            
// server-initiated reload as user activity.                                                            
export let new_user_input = true;                                                                       
```

Let's definitely dig deeper into how to fix that reproducer -- I think it's one of our more important "we've never been able to figure out why this happens" bugs, so fixing it would be amazing. 

@zbenjamin do you by any chance user Firefox? I think this might explain the unread eating behavior you mentioned a few months back that we couldn't figure out.

laurynmm replied:
So, we've got #26299 up for a page reload marking first message as read, which is a variation on this reproducer. 

Here's the conversation about that issue on CZO: https://chat.zulip.org/#narrow/stream/6-frontend/topic/messages.20marked.20as.20read.20in.20open.20tabs.20when.20page.20reloaded/near/1611162.

I think that it's related to the refactoring we did in #25387, specifically commit 8f1c3a0fa1e81e5790460e39ab261c8016cea0d2. It would be good to get your thoughts @timabbott in that conversation/issue.

timabbott commented on web/src/server_events.js:
```diff
@@ -264,6 +265,14 @@ export function force_get_events() {
+    // If user has the all messages (aka home) view as their default
+    // and current view, call select_id on current selected message
+    // again to trigger mark_read event for selected message.
+    if (!recent_topics_util.is_visible() && message_lists.current === message_lists.home) {
```
This call should I think be to `narrow_state.is_message_feed_visible` - it has the same effect today, but when we add an inbox view, the `recent_topics_util` version will become wrong.

laurynmm replied:
Done.

timabbott commented on web/src/server_events.js:
```diff
@@ -264,6 +265,14 @@ export function force_get_events() {
+    if (!recent_topics_util.is_visible() && message_lists.current === message_lists.home) {
+        const current_selected_message = message_lists.home.selected_id();
+        message_lists.home.select_id(current_selected_message);
+    }
```
What's the thinking behind putting this hook here that will mark things as read? As I understand it, `home_view_loaded` is about our having the data to fully render the home view all the way to the present ... not about some minimal level required to show it at all.

laurynmm replied:
Hmm... yeah, it's probably better to add this hook in `message_fetch.js`. Before where we call `home_view_loaded` perhaps?

The idea is that now that we're fixing the bug noted above by passing `mark_read: false` to the home message list when we initially set the selected message, we need another `select_id` call for the home message list (if it's our current view on loading) to potentially mark the selected message as read (depending on the unread state and the user's settings).

laurynmm replied:
Moved this to `message_fetch.initialize` as part of the `load_more` function.

laurynmm commented:
**Updates**:
- In the first commit:
  - Moved the hook for calling `select_id` with `mark_read: true` for `message_list.home` to still happen in the `load_more` function in `message_fetch.initialize`, but later in the loading process.
- In the second commit:
  - Added a `process_selected_message` function to be called after `process_visible` when the window gains focus to cover the case where the bottom of the feed isn't visible, but the selected message should be marked as read.
  - Both `process_visible` and `process_selected_message` here do not cover the case where a user has a modal or overlay open and the window out of focus. In this case, when the window is focused, the messages are not marked as read and remain unread when the modal or overlay is closed.

timabbott commented on web/src/message_scroll.js:
```diff
@@ -163,21 +163,23 @@ export function initialize() {
-                message_lists.current === event.msg_list
-            ) {
-                unread_ui.notify_messages_remain_unread();
+            if (unread_ops.is_window_focused()) {
```
Should this instead be making the conditional above be `if event.mark_read && unread_ops.is_window_focused()`, to avoid unnecessary extra nesting?

timabbott replied:
I guess my other potential concern about this change is that if you do some action (like say hit `pagedown`) and then immediately hit alt-tab quickly to defocus the window before the scroll finishes processing, then I think we will no longer mark messages as read? And the new `process_selected_message` logic seems like it will only do the final selected message, not the N that you scrolled past.

I wonder whether we instead want logic of the form "store the set of messages that were scrolled past" in an else clause here, and then instead of the new `process_selected_message` call, mark-as-read the set of messages that had been stored for such adjustment when the window is focused again.

Not sure, I certainly would love to avoid that sort of action-at-a-distance complexity. An alternative approach would be to instead have the check here be something like "If the window has been focused within the last 60 seconds" rather than whether it's focused right now.

zbenjamin commented:
i only use the Zulip desktop app and mobile app

May you be well,
Zev

On Wed, Jul 26 2023 at 03:31, Lauryn Menard < ***@***.*** > wrote:

> 
> 
> 
> 
> 
> 
> 
> ***@***.**** commented on this pull request.
> 
> 
> 
> In web/src/message_fetch.js (
> https://github.com/zulip/zulip/pull/26023#discussion_r1274758633 ) :
> 
> >              message_lists.home.select_id(data.anchor, {
> +              
> mark_read: false,
> 
> 
> So, we've got #26299 ( https://github.com/zulip/zulip/issues/26299 ) up for
> a page reload marking first message as read, which is a variation on this
> reproducer.
> 
> 
> 
> Here's the conversation about that issue on CZO: https://chat.zulip.org/#narrow/stream/6-frontend/topic/messages.20marked.20as.20read.20in.20open.20tabs.20when.20page.20reloaded/near/1611162
> .
> 
> 
> 
> I think that it's related to the refactoring we did in #25387 (
> https://github.com/zulip/zulip/pull/25387 ) , specifically commit 8f1c3a0 (
> https://github.com/zulip/zulip/commit/8f1c3a0fa1e81e5790460e39ab261c8016cea0d2
> ). It would be good to get your thoughts @timabbott (
> https://github.com/timabbott ) in that conversation/issue.
> 
> 
> 
> —
> Reply to this email directly, view it on GitHub (
> https://github.com/zulip/zulip/pull/26023#discussion_r1274758633 ) , or unsubscribe
> (
> https://github.com/notifications/unsubscribe-auth/AAAOKPBTIP4ZPBZQEUEFIBTXSDWXBANCNFSM6AAAAAAZGWOMVU
> ).
> You are receiving this because you were mentioned. Message ID: <zulip/zulip/pull/26023/review/1547410868
> @ github. com>
> 
> 
>

amanagr commented:
Closing as we implemented a different change in #31890 that fixed the issue.
